privacy/scrubber.go

```diff
@@ -2,251 +2,130 @@ package privacy
-	bidRequest.User = userCopy
-	return bidRequest
+	reqWrapper.SetImp(impWrapper)
 }
 
-func (s scrubber) ScrubDevice(device *openrtb2.Device, id ScrubStrategyDeviceID, ipv4 ScrubStrategyIPV4, ipv6 ScrubStrategyIPV6, geo ScrubStrategyGeo) *openrtb2.Device {
-	if device == nil {
-		return nil
-	}
-
-	deviceCopy := *device
-
-	switch id {
-	case ScrubStrategyDeviceIDAll:
-		deviceCopy.DIDMD5 = ""
-		deviceCopy.DIDSHA1 = ""
-		deviceCopy.DPIDMD5 = ""
-		deviceCopy.DPIDSHA1 = ""
-		deviceCopy.IFA = ""
-		deviceCopy.MACMD5 = ""
-		deviceCopy.MACSHA1 = ""
+func ScrubGEO(reqWrapper *openrtb_ext.RequestWrapper) {
+	//round user's geographic location by rounding off IP address and lat/lng data.
+	//this applies to both device.geo and user.geo
+	if reqWrapper.User != nil && reqWrapper.User.Geo != nil {
+		reqWrapper.User.Geo = scrubGeoPrecision(reqWrapper.User.Geo)
 	}
 
-	switch ipv4 {
-	case ScrubStrategyIPV4Subnet:
-		deviceCopy.IP = scrubIP(device.IP, s.ipV4.AnonKeepBits, iputil.IPv4BitSize)
+	if reqWrapper.Device != nil {
+		if reqWrapper.Device.Geo != nil {
+			reqWrapper.Device.Geo = scrubGeoPrecision(reqWrapper.Device.Geo)
+		}
 	}
+}
 
-	switch ipv6 {
-	case ScrubStrategyIPV6Subnet:
-		deviceCopy.IPv6 = scrubIP(device.IPv6, s.ipV6.AnonKeepBits, iputil.IPv6BitSize)
+func ScrubGeoFull(reqWrapper *openrtb_ext.RequestWrapper) {
+	if reqWrapper.User != nil && reqWrapper.User.Geo != nil {
+		reqWrapper.User.Geo = &openrtb2.Geo{}
 	}
-
-	switch geo {
-	case ScrubStrategyGeoFull:
-		deviceCopy.Geo = scrubGeoFull(device.Geo)
-	case ScrubStrategyGeoReducedPrecision:
-		deviceCopy.Geo = scrubGeoPrecision(device.Geo)
+	if reqWrapper.Device != nil && reqWrapper.Device.Geo != nil {
+		reqWrapper.Device.Geo = &openrtb2.Geo{}
 	}
 
-	return &deviceCopy
 }
 
-func (scrubber) ScrubUser(user *openrtb2.User, strategy ScrubStrategyUser, geo ScrubStrategyGeo) *openrtb2.User {
-	if user == nil {
-		return nil
+func ScrubDeviceIP(reqWrapper *openrtb_ext.RequestWrapper, ipConf IPConf) {
+	if reqWrapper.Device != nil {
+		reqWrapper.Device.IP = scrubIP(reqWrapper.Device.IP, ipConf.IPV4.AnonKeepBits, iputil.IPv4BitSize)
+		reqWrapper.Device.IPv6 = scrubIP(reqWrapper.Device.IPv6, ipConf.IPV6.AnonKeepBits, iputil.IPv6BitSize)
 	}
+}
 
-	userCopy := *user
-
-	if strategy == ScrubStrategyUserIDAndDemographic {
-		userCopy.BuyerUID = ""
-		userCopy.ID = ""
-		userCopy.Ext = scrubExtIDs(userCopy.Ext, "eids")
-		userCopy.Yob = 0
-		userCopy.Gender = ""
-	}
+func ScrubDeviceIDsIPsUserDemoExt(reqWrapper *openrtb_ext.RequestWrapper, ipConf IPConf, fieldName string, scrubFullGeo bool) {
+	ScrubDeviceIDs(reqWrapper)
+	ScrubDeviceIP(reqWrapper, ipConf)
+	ScrubUserDemographics(reqWrapper)
+	ScrubUserExt(reqWrapper, fieldName)
 
-	switch geo {
-	case ScrubStrategyGeoFull:
-		userCopy.Geo = scrubGeoFull(user.Geo)
-	case ScrubStrategyGeoReducedPrecision:
-		userCopy.Geo = scrubGeoPrecision(user.Geo)
+	if scrubFullGeo {
+		ScrubGeoFull(reqWrapper)
+	} else {
+		ScrubGEO(reqWrapper)
```

Comment:
I noticed a lack of test coverage here. Is this out of the scope of this PR to update this?

Reply:
More unit tests are added. Consolidated scrubber functions like `ScrubUserFPD` are covered by integration tests in `utils_test.go` -> `TestCleanOpenRTBRequestsActivities`